kolibri/content/utils/annotation.py

```diff
@@ -1,38 +1,163 @@
+        files = bridge.session.query(LocalFileClass).all()
+    elif type(checksums) == list:
+        logging.info('Setting availability of {number} LocalFile objects based on disk availability'.format(number=len(checksums)))
+        files = bridge.session.query(LocalFileClass).filter(LocalFileClass.id.in_(checksums)).all()
+    else:
+        logging.info('Setting availability of LocalFile object with checksum {checksum} based on disk availability'.format(checksum=checksums))
+        files = [bridge.session.query(LocalFileClass).get(checksums)]
+
+    rows_unflushed = 0
+    files_to_update = []
+    for file in files:
+        if os.path.exists(get_content_storage_file_path(get_content_file_name(file))):
+            files_to_update.append({
+                'id': file.id,
+                'available': True,
+            })
+            rows_unflushed += 1
+        if rows_unflushed == 10000:
+            bridge.session.bulk_update_mappings(LocalFileClass, files_to_update)
+            bridge.session.flush()
+            files_to_update = []
+            rows_unflushed = 0
+
+    if files_to_update:
+        bridge.session.bulk_update_mappings(LocalFileClass, files_to_update)
+
+    bridge.session.commit()
+
+    bridge.end()
+
+def recurse_availability_up_tree():
+    bridge = Bridge(app_name=CONTENT_APP_NAME)
+
+    ContentNodeClass = bridge.get_class(ContentNode)
+
+    ContentNodeTable = bridge.get_table(ContentNode)
+
+    connection = bridge.get_connection()
+
+    node_levels = bridge.session.query(ContentNodeClass.level).distinct().all()
+
+    # We can ignore the top level, as we annotate based on the level below
+    levels = sorted([level[0] for level in node_levels])[1:]
+
+    logging.info('Setting availability of ContentNode objects with children for {levels} levels'.format(levels=len(levels)))
+
+    # Go from the deepest level to the shallowest
+    for level in reversed(levels):
+        select_parents_of_available = select([ContentNodeTable.c.parent_id]).where(
+            and_(
+                ContentNodeTable.c.available == True,  # noqa
+                ContentNodeTable.c.level == level,
+            )
+        )
+        logging.info('Setting availability of ContentNode objects with children for level {level}'.format(level=level))
+        # Only modify topic availability here
+        connection.execute(ContentNodeTable.update().where(
+            ContentNodeTable.c.id.in_(select_parents_of_available)).where(
+            ContentNodeTable.c.level == level - 1).where(
+            ContentNodeTable.c.kind == content_kinds.TOPIC).values(available=True).execution_options(autocommit=True))
+        connection.execute(ContentNodeTable.update().where(
+            ContentNodeTable.c.id.notin_(select_parents_of_available)).where(
+            ContentNodeTable.c.level == level - 1).where(
+            ContentNodeTable.c.kind == content_kinds.TOPIC).values(available=False).execution_options(autocommit=True))
```

Comment:
Yeah, I also tried doing a `CASE` clause to conditionally set the availability, rather than having to execute a second statement, so I definitely think there are things we could do.

Similar concerns about number of rows being updated apply here as well.